samd: A set of small changes and bug fixes.

robert-hh opened the pull request:
- Make some machine devices configurable by #defines. These include ADC, DAC, I2C, SoftI2C, SPI, SoftI2C, PWM, UART, pulse. 
- Drop support for SoftSPI max speed. Saves ~140 bytes and is not really needed, since for higher baud rates hard SPI is available.
- Update pins.csv for the Mini SAM M4 board.
- Rearrange the mcu specific loader files.  Such that they are easier to adapt. The maximum code size is set by:

    MICROPY_HW_CODESIZE=xxxK
    
    in mpconfigmcu.mk for the MCU family as default or  in mpconfigboard.mk for a specific board. Setting the maximum code size allows the loader to raise an error if the code gets larger than the space dedicated to it.
- Add default deploy instructions.
- Enable MICROPY_ENABLE_EMERGENCY_EXCEPTION_BUF for SAMD51.
- Fix a soft reset bug, which caused soft reset to stall if streams using e.g. UART or SPI had to be closed by it's finalizer.
- Enable machine.deepsleep() as alias to machine.lightsleep(), providing consistency at least for the method set.
- Rename two pins of the WIO terminal. The names started with a digit.
- Reduce the startup time. At boards with a crystal, the startup wait time after power-up is reduced to 1 second. At boards using USB-sync, the code checks for USB registering instead of just waiting a fixed time.

robert-hh committed:
samd: Rearrange the MCU-specific loader files.

Such that they are easier to adapt.  The maximum code size is set by:

    MICROPY_HW_CODESIZE=xxxK

in mpconfigmcu.mk for the MCU family as default or in mpconfigboard.mk for
a specific board.  Setting the maximum code size allows the loader to error
out if the code gets larger than the space dedicated for it.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/modmachine: Make some machine classes configurable by #defines.

These include ADC, DAC, I2C, SoftI2C, SPI, SoftI2C, PWM, UART, pulse.  This
is useful for devices like the Adafruit Trinket series which have almost no
accessible GPIO pins.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/mpconfigport: Drop support for SoftSPI max speed.

Saves ~140 bytes and is not really needed, since for higher baud rates hard
SPI is available.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/boards/MINISAM_M4: Update pins.csv for the Mini SAM M4 board.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/mcu/samd51: Enable MICROPY_ENABLE_EMERGENCY_EXCEPTION_BUF.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/boards: Add default deploy instructions.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/main: Fix sercom deinit ordering in soft-reset.

If sockets were open when calling soft reset, gc_sweep_all() would try to
close them.  In case of e.g. the NINA WiFi handler, connected through SPI,
spi_transfer() would be called for command exchange with the NINA module.
But at that time SerCom was already disabled.

Moving sercom_deinit_all() behind gc_sweep_all() solves this issue.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/modmachine: Add machine.deepsleep as alias of machine.lightsleep.

Just for coverage.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented on ports/samd/Makefile:
```diff
@@ -175,8 +178,10 @@ endif
 
 $(BUILD)/firmware.elf: $(OBJ)
+	$(ECHO) "PREPROCESS LINK $@"
+	$(Q)$(CC) -E -x c $(LDDEFINES) $(LD_FILES)| grep -v '^#' > $(BUILD)/link.ld
```
Instead of using the C preprocessor, it's possible to define symbols using `--defsym`.  Eg stm32 does:
```
LDFLAGS += -Wl,--defsym=_estack_reserve=8
```
The nrf port also uses `defsym`.  I think it would be simpler to use that instead of CPP.

robert-hh replied:
I'll try that. Using the preprocessor was carried over from the MIMXRT port, but it's ugly. If using symdef works, it is more straight.

robert-hh committed:
samd/boards/SEEED_WIO_TERMINAL: Rename two pins starting with a digit.

So the names don't start with a digit, which isn't allowed in Python.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/mcu: Reduce the startup time after hard reset.

With Crystal: set the crystal startup wait time to 1 second.  It was 2
seconds before, and that seeemed too long.

With USB-Sync: scan for up to 1 second for the USB to be registered and
carry on with boot as soon as it it.  Before, the code just waited for
500ms.

Side change: improve related comments.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented:
Thanks for updating.  This is now merged.

robert-hh commented:
Thank you very much. Next in queue for SAMD may be #10233, which helps. It increases both flash write speed and code execution speed at SAMD51 devices.
